crypto/rsa/rsa_lib.c

```diff
@@ -906,6 +906,57 @@ int ossl_rsa_get0_all_params(RSA *r, STACK_OF(BIGNUM_const) *primes,
     return 1;
 }
 
+#define safe_BN_num_bits(_k_)           (((_k_) == NULL) ? 0 : BN_num_bits((_k_)))
+int ossl_rsa_check_factors(RSA *r)
+{
+    int valid = 0;
+    int n, d_ok, factors_ok, exps_ok, coeffs_ok, bits, i;
+    STACK_OF(BIGNUM_const) *factors = sk_BIGNUM_const_new_null();
+    STACK_OF(BIGNUM_const) *exps = sk_BIGNUM_const_new_null();
+    STACK_OF(BIGNUM_const) *coeffs = sk_BIGNUM_const_new_null();
+
+    if (factors == NULL || exps == NULL || coeffs == NULL)
+        goto done;
+
+    /*
+     * Simple sanity check for RSA key. All RSA key parameters
+     * must be less-than/equal-to RSA parameter n
+     */
+    ossl_rsa_get0_all_params(r, factors, exps, coeffs);
+    n = safe_BN_num_bits(RSA_get0_n(r));
+
+    exps_ok = 1;
+    for (i = 0; i < sk_BIGNUM_const_num(exps); i++) {
+        bits = safe_BN_num_bits(sk_BIGNUM_const_value(exps, i));
+        exps_ok = ((exps_ok) && (bits <= n));
+    }
+
+    d_ok = (safe_BN_num_bits(RSA_get0_d(r)) <= n);
+
+    n = (n / 2) + 2;	/* ? rounding ? Avoids failures in evp_test */
```

Comment:
we only know p*q = N
but why can you be so sure that there must be an upper bound
for the bitsize of p and q, except n ?

Reply:
I'm not sure with this. Originally I took what's been suggested [here](https://github.com/openssl/openssl/pull/25190#discussion_r1718096292). And got some failures. After investigating those failures I saw `n/2`
is too strict (wrong).

I'm trying to figure out a better check. So perhaps with your hint here I should be rather doing:
`safe_BN_num_bits(q) + safe_BN_num_bis(q) == n` ?

thanks.

Reply:
The only mathematical rule I can see is 2<p<N/2 and 2<q<N/2,
and I don't see what the benefit of guessing slightly more restrictive
limits is, with regard to the issue at hand.
PS: your suggested equation is already violated by p=q=3 => N=9,
num_bits(p)+num_bits(q)=2+2 != num_bits(9)=3

Reply:
> we only know p*q = N but why can you be so sure that there must be an upper bound for the bitsize of p and q, except n ?

I always thought that *p* and *q* should be roughly equally sized, that's where I cam from when proposing that.  On the other hand, having either being (almost) the same size as *N* would be absurd...

Reply:
> I'm trying to figure out a better check. So perhaps with your hint here I should be rather doing: `safe_BN_num_bits(q) + safe_BN_num_bis(q) == n` ?

Hopefully with one of the operands being safe_BN_num_bits(*p*) 😉

Reply:
sure, but we just have to avoid a DoS attack due to excessive time spend in checking
the bogus prime factors.

Reply:
just to clarify:
> 2<p<N/2 and 2<q<N/2
talks about values of `p`, `q` and `N`. Not number of bits, is that correct?

Reply:
yes, p and q are the prime factors of N, and n means log2(N)

Reply:
Roughly speaking, ok pseudocode, but I find the `fail` logic hard to read.  How about:

```
for (i = 0, ok = 1; i < factors; i++)
   ok = ok && (safe_BN_num_bits(q[i]) + safe_BN_num_bis(p[i])) != n;
```

...  also, if constant time isn't a concern, you could short-circuit like this:

```
for (i = 0; i < factors; i++)
   if ((safe_BN_num_bits(q[i]) + safe_BN_num_bis(p[i])) != n)
      break;
ok = ok && i == factors;
```

Reply:
In case of a 3-prime RSA, we would expect that
p*q*r = N,
and p, q, and r having around n/3 bits,
but the exact limits depend on how the prime factors
are generated.

Reply:
`safe_BN_num_bits(q[i]) + safe_BN_num_bis(p[i]) != n`
no sorry, this wont work...

Reply:
I really do not think this test must (or even should) be precise. So even comparing with <= n bitsize would be sufficient. Please note this is only a very rough sanity check.

Reply:
We have EVP_PKEY_check*() calls for real key validations.